### PR TITLE
made QuerySet iteration respect chunk_size

### PR DESCRIPTION
**Pull request by timgraham**

### CODE REVIEW

File: django_mongodb/compiler.py [hunk 1]
```diff
@@ -33,11 +35,21 @@ def execute_sql(
         except EmptyResultSet:
             return iter([]) if result_type == MULTI else None
 
-        return (
-            (self._make_result(row, columns) for row in query.fetch())
-            if result_type == MULTI
-            else self._make_result(next(query.fetch()), columns)
-        )
+        cursor = query.get_cursor()
+        if result_type == SINGLE:
```
Comment (timgraham) [inline django_mongodb/compiler.py:39]: @WaVEV I pulled in the `if result_type == SINGLE:` block from https://github.com/mongodb-labs/django-mongodb/pull/84. Please check this patch doesn't break anything for you there. This block isn't used until that PR.

Comment (WaVEV) [reply django_mongodb/compiler.py:39]: it seems equivalent.

File: django_mongodb/compiler.py [hunk 2]
```diff
@@ -86,6 +107,15 @@ def _make_result(self, entity, columns):
             result.append(obj.get(name))
         return result
 
+    def cursor_iter(self, cursor, chunk_size, columns):
+        """Generator to yield chunks from cursor."""
+        chunk = []
+        for i, row in enumerate(cursor):
+            if i % chunk_size == 0 and i > 0:
+                yield chunk
```
Comment (WaVEV) [inline django_mongodb/compiler.py:115]: possible bug, always happens. Shouldn't chunk get re-initialized here?

Comment (WaVEV) [reply django_mongodb/compiler.py:115]: the other thing, you could move line 116 up above the if and you can remove the `i > 0` check

Comment (WaVEV) [reply django_mongodb/compiler.py:115]: And, the last query. 😄 . you aren't using `len(chunk)` I think the function would be easy to red in the following way:
```python
def cursor_iter(self, cursor, chunk_size, columns):
    """Generator to yield chunks from cursor."""
    chunk = []
    for row in cursor:
        if len(chunk) == chunk_size:
            yield chunk
            chunk = []
    yield chunk
```

**Review by WaVEV — APPROVED** [review]

File: django_mongodb/compiler.py [hunk 3]
```diff
@@ -33,11 +35,21 @@ def execute_sql(
...
+                return self._make_result(obj, columns)
+        # result_type is MULTI
+        cursor.batch_size(chunk_size)
+        result = self.cursor_iter(cursor, chunk_size, columns)
+        if not chunked_fetch:
+            # If using non-chunked reads, read data into memory.
+            return list(result)
+        return result
 
     def results_iter(
```
Comment (Jibola) [inline django_mongodb/compiler.py:54]: Sorry if this is a dumb question. How is this getting called in relation to `_make_result` or `cursor_iter`

File: django_mongodb/compiler.py [hunk 4]
```diff
@@ -86,6 +107,16 @@ def _make_result(self, entity, columns):
...
         return result
 
+    def cursor_iter(self, cursor, chunk_size, columns):
+        """Yield chunks of results from cursor."""
+        chunk = []
+        for row in cursor:
+            chunk.append(self._make_result(row, columns))
+            if len(chunk) == chunk_size:
+                yield chunk
+                chunk = []
```
Comment (Jibola) [inline django_mongodb/compiler.py:117]: So "Chunk" here is the number of items in a list and not bounded by bytes? (confirming as that was the expectation)

Comment (timgraham) [reply django_mongodb/compiler.py:117]: Right. Django's default for `chunk_size` is 100 (rows), or 2000 if using `QuerySet.iterator()`.

File: django_mongodb/compiler.py (hunk 3 above)
Comment (timgraham) [reply django_mongodb/compiler.py:54]: It's called afterward, except when `results is None` and this method invokes `execute_sql()` itself. You can look at `ModelIterable.__iter__()` in [django/db/models/query.py](https://github.com/django/django/blob/main/django/db/models/query.py) and see `compiler.execute_sql` and `compiler.results_iter`

**Pull request merged**